File: Olympiads/CCC/CCC '02 S3 - Blindfold.py
r = int(input())
c = int(input())

di = [-1,0,1,0]
dj = [0,1,0,-1]

grid = []

for i in range(r):
    grid.append(list(input()))

n = int(input())

t = []

for i in range(n):
    t.append(input())

last = []
offset = []

for i in range(4):
    offset.append(set())
    



for i in range(4):
    x = 0
    y = 0
    dir = i

    offset[i].add((x,y))

    for j in t:
        
        if(j == "F"):
            x += di[dir]
            y += dj[dir]
        
        if(j == "L"):
            dir = (dir+4-1) % 4
        if(j == "R"):
            dir = (dir+4+1)%4
        
        offset[i].add((x,y))
        
    
    last.append((x,y))

#print(last[0][0], last[0][1])





for i in range(r):
    for j in range(c):
        if grid[i][j] == "X": 
            continue

        for k in range(4):
            flag = True
            for a,b in offset[k]:
                
                if i + a < 0 or i + a >= r or j + b < 0 or j + b >= c:
                    flag = False
                    break
                if grid[i+a][j+b] == "X":
                    
                    flag = False
                    break

                
            
            if(flag):
                
                grid[i+last[k][0]][j+last[k][1]] = "*"

for i in range(r):
    print("".join(grid[i]))

File: Olympiads/CCC/CCC '02 S3 - Blindfold.cpp
#include <bits/stdc++.h>
#include <iostream>

using namespace std;

typedef pair<int,int> pii;

const int MAXN = 3e4+5;

int n, r, c, t[MAXN];

vector<pii> offset[4];

//bool visited[400][85][4][MAXN];

char grid[400][85];

vector<int> di = {-1,0,1,0}, dj = {0,1,0,-1};


int main(){

    ios_base::sync_with_stdio(false);
    cin.tie(NULL);


    cin >> r >> c;

    for(int i = 0; i < r; i++){
        string s; cin >> s;

        for(int j = 0; j < c; j++){
            grid[i][j] = s[j];
            
        }
    }

    cin >> n;

    


    for(int i = 0; i < n; i++){
        char s; cin >> s;

        if(s == 'F') t[i] = 0;
        if(s == 'L') t[i] = -1;
        if(s == 'R') t[i] = 1;
    }

    for(int i = 0; i < 4; i++){
        int x = 0;
        int y = 0;
        int dir = i;

      

        for(int j = 0; j < n; j++){
            if(!t[j]){
                x += di[dir];
                y += dj[dir];
            }
            else{
                dir = (dir+4+t[j]) % 4;
            }

            offset[i].push_back({x,y});


        }
    }
    /*
    for(int i = 0; i < 4; i++){
        
        sort(offset[i].begin(), offset[i].end());
        offset[i].erase(unique(offset[i].begin(), offset[i].end() ), offset[i].end());
        
    
    }
    */
    

    //0 is up
    //1 is right
    //2 is down
    //3 is left

   

    for(int i = 0; i < r; i++){
        for(int j = 0; j < c; j++){

            if(grid[i][j] == 'X') continue;
            for(int k = 0; k < 4; k++){

                bool check = 1;
                for(pii o : offset[k]){
                    if(i + o.first < 0 || i+o.first >= r || j + o.second < 0 || j + o.second >= c){
                        check = 0;
                        break;
                    }

                    if(grid[i+o.first][j+o.second] == 'X'){
                        check = 0;
                        break;
                    }
                }

                if(check){
                    //cout << i << " " << offset[k][n-1].first << " " << j << " " << offset[k][n-1].second << " " << k << "\n";
                    grid[i+offset[k][n-1].first][j+offset[k][n-1].second] = '*';
                }
            }
        }
    }

    for(int i = 0; i < r; i++){
        for(int j = 0; j < c; j++){
            cout << grid[i][j];
        }

        cout << "\n";
    }
    
    return 0;

    



    

    

    /*
    for(int i = 0; i < r; i++){
        for(int j = 0; j < c; j++){
            if(grid[i][j] == 'X') break;
            for(int l = 0; l < 4; l++){
                int x = i;
                int y = j;
                bool check = 1;
                int dir = l;

                

                for(int k = 0; k < n; k++){
                    
                    if(t[k] == 0){
                        x += di[dir];
                        y += dj[dir];
                    }

                    else{
                        dir = (dir+4+t[k]) % 4;
                    }

                    //if(visited[x][y][dir][k]){
                        //check = 0;
                        //break;
                    //}

                    //visited[x][y][dir][k] = 1;

                    //if(i == 0 && j == 0 && l == 1) cout << x << " " << y << " " << dir << " " << grid[x][y] << "\n";
                    

                    if(x < 0 || x >= r || y < 0 || y >= c){
                        //cout << x << " " << y << " " << dir << " " << k << "\n";
                        check = 0;
                        break;
                    }

                    if(grid[x][y] == 'X'){
                        //cout << x << " " << y << " " << dir << " " << k << "\n";
                        check = 0;
                        break;
                    }

                }

                

                if(check){
                    
                    grid[x][y] = '*';
                } 
                




            }
        }
    }

    */

    

    
    
}